### PR TITLE
feat(headless/commerce): rework cart controller

### PR DESCRIPTION
**Pull request by fbeaudoincoveo**
https://coveord.atlassian.net/browse/CAPI-191

### CODE REVIEW

File: packages/headless/src/commerce.index.ts [hunk 1]
```diff
@@ -71,7 +71,6 @@ export type {
   CartProps,
   Cart,
   CartState,
-  CartControllerState,
```
Comment (Spuffynism) [inline packages/headless/src/commerce.index.ts:74]: Why do we not expose this anymore?

File: packages/headless/src/controllers/commerce/context/cart/headless-cart.test.ts [hunk 2]
```diff
@@ -1,87 +1,140 @@
...
+  });
+
+  it('exposes a #subscribe method', () => {
+    expect(cart.subscribe).toBeTruthy();
+  });
 
   it('adds the correct reducers to engine', () => {
     expect(engine.addReducers).toHaveBeenCalledWith({
       cart: cartReducer,
+      context: contextReducer,
```
Comment (Spuffynism) [inline packages/headless/src/controllers/commerce/context/cart/headless-cart.test.ts:65]: Why do we load the context reducer?

If someone uses a cart controller, wouldn't they have already loaded a context controller?

File: packages/headless/src/controllers/commerce/context/cart/headless-cart.test.ts [hunk 3]
```diff
@@ -1,87 +1,140 @@
...
+
+    cart.empty();
+
+    const item = options.items![0];
+
+    expect(updateItemSpy).toHaveBeenCalledWith({
+      ...item,
+      quantity: 0,
+    });
   });
```
Comment (Spuffynism) [inline packages/headless/src/controllers/commerce/context/cart/headless-cart.test.ts:99]: We _could_ group these tests + the `initializes` test case under an `initialization` `describe` group.

File: packages/headless/src/controllers/commerce/context/cart/headless-cart.ts [hunk 4]
```diff
@@ -35,41 +60,65 @@ export interface CartProps {
...
   /**
-   * Replaces the cart items.
-   * @param items - The new cart items.
-   */
-  setItems(items: CartItem[]): void;
-
-  /**
-   * Adds a cart item.
-   * @param item - The new cart item.
+   * Creates, updates, or deletes an item in the cart, and emits an `ec.cartAction` analytics event if the `quantity` of
```
Comment (Spuffynism) [inline packages/headless/src/controllers/commerce/context/cart/headless-cart.ts:63]: Really nice work on this doc! ⭐

Comment (Spuffynism) [inline packages/headless/src/controllers/commerce/context/cart/headless-cart.ts:63]: I remember seeing an RFC on naming events with a `ec.` vs `ec_` prefix. Has this been settled?

File: packages/headless/src/controllers/commerce/context/cart/headless-cart.ts [hunk 5]
```diff
@@ -97,35 +146,52 @@ export function buildCart(engine: CommerceEngine, props: CartProps = {}): Cart {
...
-        cart: cartItems.map((id: string) => cart[id]) as CartItem[],
-      };
+    empty: function () {
+      for (const item of itemsSelector(getState())) {
+        this.updateItem({...item, quantity: 0});
+      }
+    },
+
+    purchase: (_transactionId: string, _transactionRevenue: number) => {
+      // TODO: log ec.purchase with all products in cart.
```
Comment (Spuffynism) [inline packages/headless/src/controllers/commerce/context/cart/headless-cart.ts:168]: We can map this TODO to https://coveord.atlassian.net/browse/LENS-1498

File: packages/headless/src/controllers/commerce/context/cart/headless-cart.ts [hunk 6]
```diff
@@ -97,35 +146,52 @@ export function buildCart(engine: CommerceEngine, props: CartProps = {}): Cart {
...
+      dispatch(setItems([]));
     },
 
-    setItems: (items: CartItem[]) => dispatch(setItems(items)),
+    updateItem(item: CartItem) {
+      if (isNewQuantityDifferent(item.productId, item.quantity)) {
+        // TODO: log ec.cartAction; if new quantity > previous, 'add', otherwise, 'remove'.
```
Comment (Spuffynism) [inline packages/headless/src/controllers/commerce/context/cart/headless-cart.ts:174]: We can map this TODO to https://coveord.atlassian.net/browse/LENS-1497

File: packages/headless/src/controllers/commerce/context/cart/headless-cart.ts [hunk 7]
```diff
@@ -97,35 +146,52 @@ export function buildCart(engine: CommerceEngine, props: CartProps = {}): Cart {
...
-      return {
-        cart: cartItems.map((id: string) => cart[id]) as CartItem[],
-      };
+    empty: function () {
+      for (const item of itemsSelector(getState())) {
+        this.updateItem({...item, quantity: 0});
+      }
+    },
+
+    purchase: (_transactionId: string, _transactionRevenue: number) => {
```
Comment (Spuffynism) [inline packages/headless/src/controllers/commerce/context/cart/headless-cart.ts:167]: Could we define this method with the same syntax as the one defined below?

**Review by Spuffynism — COMMENTED** [review]
Nice work!

File: packages/headless/src/controllers/commerce/context/cart/headless-cart.ts (hunk 4 above)
Comment (fbeaudoincoveo) [reply packages/headless/src/controllers/commerce/context/cart/headless-cart.ts:63]: Not that I'm aware of, but we'll update the doc if needed 😄

File: packages/headless/src/commerce.index.ts (hunk 1 above)
Comment (fbeaudoincoveo) [reply packages/headless/src/commerce.index.ts:74]: It was defined as 

```
export type CartControllerState = Cart['state'];
```

i.e., `CartState`

So I don't really see the need for it if we export `CartState`.

File: packages/headless/src/controllers/commerce/context/cart/headless-cart.test.ts (hunk 2 above)
Comment (fbeaudoincoveo) [reply packages/headless/src/controllers/commerce/context/cart/headless-cart.test.ts]: Yep you're right, I don't think we need to load this reducer after all

**Review by louis-bompart — APPROVED** [review]

**Review by Spuffynism — APPROVED** [review]
fyi @samisayegh TODOs in this PR link to a few tasks to add relay support to headless commerce

**Comment by fbeaudoincoveo** [comment]
https://coveord.atlassian.net/browse/CAPI-191

**Pull request merged**